src/providers/wfs/qgsoapifcollection.cpp

```diff
@@ -0,0 +1,497 @@
+  // lives in the main thread without risking havoc.
+  connect( this, &QgsBaseNetworkRequest::downloadFinished, this, &QgsOapifCollectionsRequest::processReply, Qt::DirectConnection );
+}
+
+bool QgsOapifCollectionsRequest::request( bool synchronous, bool forceRefresh )
+{
+  if ( !sendGET( QUrl( mUrl ), QStringLiteral( "application/json" ), synchronous, forceRefresh ) )
+  {
+    emit gotResponse();
+    return false;
+  }
+  return true;
+}
+
+QString QgsOapifCollectionsRequest::errorMessageWithReason( const QString &reason )
+{
+  return tr( "Download of collections description failed: %1" ).arg( reason );
+}
+
+void QgsOapifCollectionsRequest::processReply()
+{
+  if ( mErrorCode != QgsBaseNetworkRequest::NoError )
+  {
+    emit gotResponse();
+    return;
+  }
+  const QByteArray &buffer = mResponse;
+  if ( buffer.isEmpty() )
+  {
+    mErrorMessage = tr( "empty response" );
+    mErrorCode = QgsBaseNetworkRequest::ServerExceptionError;
+    emit gotResponse();
+    return;
+  }
+
+  QgsDebugMsgLevel( QStringLiteral( "parsing collections response: " ) + buffer, 4 );
+
+  QTextCodec::ConverterState state;
+  QTextCodec *codec = QTextCodec::codecForName( "UTF-8" );
+  Q_ASSERT( codec );
+
+  const QString utf8Text = codec->toUnicode( buffer.constData(), buffer.size(), &state );
+  if ( state.invalidChars != 0 )
+  {
+    mErrorCode = QgsBaseNetworkRequest::ApplicationLevelError;
+    mAppLevelError = ApplicationLevelError::JsonError;
+    mErrorMessage = errorMessageWithReason( tr( "Invalid UTF-8 content" ) );
+    emit gotResponse();
+    return;
+  }
+
+  try
+  {
+    const json j = json::parse( utf8Text.toStdString() );
+
+    const auto links = QgsOAPIFJson::parseLinks( j );
+    QStringList licenses;
+    std::set<QString> licenseSet;
+    for ( const auto &link : links )
+    {
+      if ( link.rel == QLatin1String( "license" ) )
+      {
+        auto license =  !link.title.isEmpty() ? link.title : link.href;
+        if ( licenseSet.find( license ) == licenseSet.end() )
+        {
+          licenseSet.insert( license );
+          licenses << license;
+        }
+      }
+    }
+
+    if ( j.is_object() && j.contains( "collections" ) )
+    {
+      const auto collections = j["collections"];
+      if ( collections.is_array() )
+      {
+        for ( const auto jCollection : collections )
+        {
+          QgsOapifCollection collection;
+          if ( collection.deserialize( jCollection ) )
+          {
+            if ( collection.mLayerMetadata.licenses().isEmpty() )
+            {
+              // If there are not licenses from the collection description,
+              // use the one from the collection set.
+              collection.mLayerMetadata.setLicenses( licenses );
+            }
+            mCollections.emplace_back( collection );
+          }
+        }
+      }
+    }
+
+    // Paging informal extension used by api.planet.com/
```

Comment:
I thought that next/prev it were "permitted" by the specs. Is this an extension?

Reply:
It is not clear to me if paging of collections themselves is somewhat really foreseen by the spec. It is in the "spirit" of it, but never mentioned, contrary to paging of features